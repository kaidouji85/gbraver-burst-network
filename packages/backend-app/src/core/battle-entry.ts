import type { ArmdozerId, PilotId } from "gbraver-burst-core";

import type { UserID } from "./user";
import type { WSAPIGatewayConnectionId } from "./ws-api-gateway-connection";

/**
 * ネット対戦エントリ
 * カジュアルマッチ、プライベートマッチ等、全ケースのエントリで
 * 必要最小限の情報をあつめた
 */
export type BattleEntry = {
  /** エントリーするユーザのID */
  userID: UserID;

  /** 選択したアームドーザID */
  armdozerId: ArmdozerId;

  /** 選択したパイロットID */
  pilotId: PilotId;

  /** コネクションID */
  connectionId: WSAPIGatewayConnectionId;
};
