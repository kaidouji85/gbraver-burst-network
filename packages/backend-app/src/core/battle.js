// @flow

import type { Player } from "gbraver-burst-core";
import type { GameState } from "gbraver-burst-core/lib/state/game-state";

import type { UserID } from "./user";
import type { WSAPIGatewayConnectionId } from "./ws-api-gateway-connection";

/** バトルID */
export type BattleID = string;

/** ステートが更新されるたびに発行されるユニークID */
export type FlowID = string;

/** バトルに参加するプレイヤー情報 */
export type BattlePlayer = Player & {
  /** ユーザID */
  userID: UserID,
  /** コネクションID */
  connectionId: WSAPIGatewayConnectionId,
};

/**
 * バトル情報
 * @template X プレイヤー情報
 */
export type Battle<X: BattlePlayer> = {
  /** バトルID */
  battleID: BattleID,
  /** フローID */
  flowID: FlowID,
  /** プレイヤー情報 */
  players: [X, X],
  /**
   * バトル更新ポーリングをするユーザのID
   * playersに含まれているユーザのIDを指定すること
   */
  poller: UserID,
  /** ステートヒストリー */
  stateHistory: GameState[],
};
