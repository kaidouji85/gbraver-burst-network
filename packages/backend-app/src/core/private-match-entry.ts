import { BattleEntry } from "./battle-entry";
import { PrivateMatchRoomID } from "./private-match-room";

/** プライベートマッチエントリ */
export type PrivateMatchEntry = BattleEntry & {
  /** エントリするルームID */
  roomID: PrivateMatchRoomID;
};
