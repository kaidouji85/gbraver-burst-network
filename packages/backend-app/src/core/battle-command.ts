import { Command, CommandSchema } from "gbraver-burst-core";
import { z } from "zod";

import { BattleID, BattleIDSchema, FlowID, FlowIDSchema } from "./battle";
import { UserID, UserIDSchema } from "./user";

/** バトルコマンド */
export type BattleCommand = {
  /** ユーザID */
  userID: UserID;
  /** バトルID */
  battleID: BattleID;
  /** フローID */
  flowID: FlowID;
  /** コマンド */
  command: Command;
};

/** BattleCommand zodスキーマ */
export const BattleCommandSchema = z.object({
  userID: UserIDSchema,
  battleID: BattleIDSchema,
  flowID: FlowIDSchema,
  command: CommandSchema,
});