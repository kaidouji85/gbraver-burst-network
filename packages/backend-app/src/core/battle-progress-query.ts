import { BattleID, FlowID } from "./battle";

/** バトル進行 問合わせパラメータ */
export type BattleProgressQuery = {
  /** バトルID */
  battleID: BattleID;
  /** フローID */
  flowID: FlowID;
};