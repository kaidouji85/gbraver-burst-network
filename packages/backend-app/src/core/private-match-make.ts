import { BattleEntry } from "./battle-entry";
import { PrivateMatchEntry } from "./private-match-entry";
import { PrivateMatchRoom } from "./private-match-room";

/** マッチング結果 */
export type Matching = [BattleEntry, BattleEntry];

/**
 * プライベートマッチメイクを行う
 * マッチメイクできなかった場合はnullを返す
 * @param room ルーム
 * @param entries エントリ
 * @return マッチメイク結果
 */
export function privateMatchMake(
  room: PrivateMatchRoom,
  entries: PrivateMatchEntry[]
): Matching | null {
  if (entries.length <= 0) {
    return null;
  }

  const ownerEntry: BattleEntry = {
    userID: room.owner,
    connectionId: room.ownerConnextionId,
    armdozerId: room.armdozerId,
    pilotId: room.pilotId,
  };
  const otherEntry = entries[0];
  return [ownerEntry, otherEntry];
}
