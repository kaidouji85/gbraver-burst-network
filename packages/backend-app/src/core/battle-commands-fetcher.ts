import { BattlePlayer } from "./battle";
import { BattleCommand } from "./battle-command";

/** ゲーム参加プレイヤーのバトルコマンドを取得 */
export interface BattleCommandsFetcher {
  /**
   * ゲーム参加プレイヤーのバトルコマンドを取得する
   * @param players ゲーム参加プレイヤー
   * @return 取得結果、取得できなかった場合はnull
   */
  fetch(
    players: [BattlePlayer, BattlePlayer],
  ): Promise<[BattleCommand, BattleCommand] | null>;
}
