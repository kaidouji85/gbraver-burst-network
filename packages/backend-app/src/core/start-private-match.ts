import { Battle, BattlePlayer } from "./battle";
import { Connection } from "./connection";
import { createBattle } from "./create-battle";
import { createBattlePlayer } from "./create-battle-player";
import { notChosenPrivateMatchEntries } from "./not-chosen-private-match-entries";
import { PrivateMatchEntry } from "./private-match-entry";
import { PrivateMatching } from "./private-match-make";

/** プライベートマッチ開始情報 */
type MatchStartResponse = {
  /** バトル情報 */
  battle: Battle<BattlePlayer>;
  /** バトル参加者のコネクション */
  battleConnections: Connection[];
  /** マッチングしなかったプレイヤーのコネクション */
  notChosenConnections: Connection[];
};

/**
 * プライベートマッチを開始する
 * @param entries 指定したルームのすべてのエントリ
 * @param matching マッチング結果
 * @returns プライベートマッチ開始情報
 */
export function startPrivateMatch(
  entries: PrivateMatchEntry[],
  matching: PrivateMatching,
): MatchStartResponse {
  const battle = createBattle([
    createBattlePlayer(matching[0]),
    createBattlePlayer(matching[1]),
  ]);
  const battleConnections: Connection[] = matching.map((v) => ({
    connectionId: v.connectionId,
    userID: v.userID,
    state: {
      type: "InBattle",
      battleID: battle.battleID,
      players: battle.players,
    },
  }));
  const notChosenEntries = notChosenPrivateMatchEntries(matching, entries);
  const notChosenConnections: Connection[] = notChosenEntries.map((v) => ({
    connectionId: v.connectionId,
    userID: v.userID,
    state: {
      type: "None",
    },
  }));
  return { battle, battleConnections, notChosenConnections };
}
