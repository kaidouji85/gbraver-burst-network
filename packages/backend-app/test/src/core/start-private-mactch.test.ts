import { ArmdozerIds, PilotIds } from "gbraver-burst-core";
import { v4 as uuidv4 } from "uuid";

import { BattleEntry } from "../../../src/core/battle-entry";
import { PrivateMatchEntry } from "../../../src/core/private-match-entry";
import { PrivateMatching } from "../../../src/core/private-match-make";
import { startPrivateMatch } from "../../../src/core/start-private-match";
import { mockUniqUUID } from "../../mock-unique-uuid";

jest.mock("uuid");

/** テスト用ルームID */
const roomID = "test-room-id";

/** ルームオーナーのエントリ */
const ownerEntry: BattleEntry = {
  userID: "owner",
  connectionId: "owner-connection-id",
  armdozerId: ArmdozerIds.SHIN_BRAVER,
  pilotId: PilotIds.SHINYA,
};

/** 参加者1のエントリ */
const user1Entry: PrivateMatchEntry = {
  userID: "user1",
  connectionId: "user1-connection-id",
  armdozerId: ArmdozerIds.NEO_LANDOZER,
  pilotId: PilotIds.GAI,
  roomID,
};

/** 参加者2のエントリ */
const user2Entry: PrivateMatchEntry = {
  userID: "user2",
  connectionId: "user2-connection-id",
  armdozerId: ArmdozerIds.WING_DOZER,
  pilotId: PilotIds.TSUBASA,
  roomID,
};

/** 参加者3のエントリ */
const user3Entry: PrivateMatchEntry = {
  userID: "user3",
  connectionId: "user3-connection-id",
  armdozerId: ArmdozerIds.LIGHTNING_DOZER,
  pilotId: PilotIds.RAITO,
  roomID,
};

/** プライベートマッチルームの全エントリ */
const entries: PrivateMatchEntry[] = [user1Entry, user2Entry, user3Entry];

/** マッチング結果 */
const matching: PrivateMatching = [ownerEntry, user2Entry];

test("正しくプライベートマッチを開始できる", () => {
  (uuidv4 as jest.Mock).mockImplementation(mockUniqUUID());
  expect(startPrivateMatch(entries, matching)).toMatchSnapshot();
});
