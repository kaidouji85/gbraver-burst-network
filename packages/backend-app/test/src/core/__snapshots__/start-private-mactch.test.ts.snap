// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`正しくプライベートマッチを開始できる 1`] = `
{
  "battle": {
    "battleID": "mock-id-3",
    "flowID": "mock-id-4",
    "players": [
      {
        "armdozer": {
          "burst": {
            "recoverBattery": 5,
            "type": "RecoverBattery",
          },
          "id": "SHIN_BRAVER",
          "maxBattery": 5,
          "maxHp": 3100,
          "name": "シンブレイバー",
          "power": 2000,
          "speed": 2000,
        },
        "connectionId": "owner-connection-id",
        "pilot": {
          "id": "SHINYA",
          "name": "シンヤ",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "mock-id-1",
        "userID": "owner",
      },
      {
        "armdozer": {
          "burst": {
            "recoverBattery": 3,
            "type": "ContinuousAttack",
          },
          "id": "WING_DOZER",
          "maxBattery": 5,
          "maxHp": 2700,
          "name": "ウィングドーザ",
          "power": 2200,
          "speed": 2200,
        },
        "connectionId": "user2-connection-id",
        "pilot": {
          "id": "TSUBASA",
          "name": "ツバサ",
          "skill": {
            "batteryEnhancement": 1,
            "duration": 2,
            "type": "BatteryEnhancementSkill",
          },
        },
        "playerId": "mock-id-2",
        "userID": "user2",
      },
    ],
    "poller": "owner",
    "stateHistory": [
      {
        "activePlayerId": "mock-id-2",
        "effect": {
          "name": "StartGame",
        },
        "players": [
          {
            "armdozer": {
              "battery": 5,
              "burst": {
                "recoverBattery": 5,
                "type": "RecoverBattery",
              },
              "effects": [],
              "enableBurst": true,
              "hp": 3100,
              "id": "SHIN_BRAVER",
              "maxBattery": 5,
              "maxHp": 3100,
              "name": "シンブレイバー",
              "power": 2000,
              "speed": 2000,
            },
            "pilot": {
              "enableSkill": true,
              "id": "SHINYA",
              "name": "シンヤ",
              "skill": {
                "recoverBattery": 2,
                "type": "RecoverBatterySkill",
              },
            },
            "playerId": "mock-id-1",
          },
          {
            "armdozer": {
              "battery": 5,
              "burst": {
                "recoverBattery": 3,
                "type": "ContinuousAttack",
              },
              "effects": [],
              "enableBurst": true,
              "hp": 2700,
              "id": "WING_DOZER",
              "maxBattery": 5,
              "maxHp": 2700,
              "name": "ウィングドーザ",
              "power": 2200,
              "speed": 2200,
            },
            "pilot": {
              "enableSkill": true,
              "id": "TSUBASA",
              "name": "ツバサ",
              "skill": {
                "batteryEnhancement": 1,
                "duration": 2,
                "type": "BatteryEnhancementSkill",
              },
            },
            "playerId": "mock-id-2",
          },
        ],
      },
      {
        "activePlayerId": "mock-id-2",
        "effect": {
          "name": "InputCommand",
          "players": [
            {
              "command": [
                {
                  "battery": 0,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 1,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 2,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 3,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 4,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 5,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "type": "BURST_COMMAND",
                },
                {
                  "type": "PILOT_SKILL_COMMAND",
                },
              ],
              "playerId": "mock-id-1",
              "selectable": true,
            },
            {
              "command": [
                {
                  "battery": 0,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 1,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 2,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 3,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 4,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 5,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "type": "BURST_COMMAND",
                },
                {
                  "type": "PILOT_SKILL_COMMAND",
                },
              ],
              "playerId": "mock-id-2",
              "selectable": true,
            },
          ],
        },
        "players": [
          {
            "armdozer": {
              "battery": 5,
              "burst": {
                "recoverBattery": 5,
                "type": "RecoverBattery",
              },
              "effects": [],
              "enableBurst": true,
              "hp": 3100,
              "id": "SHIN_BRAVER",
              "maxBattery": 5,
              "maxHp": 3100,
              "name": "シンブレイバー",
              "power": 2000,
              "speed": 2000,
            },
            "pilot": {
              "enableSkill": true,
              "id": "SHINYA",
              "name": "シンヤ",
              "skill": {
                "recoverBattery": 2,
                "type": "RecoverBatterySkill",
              },
            },
            "playerId": "mock-id-1",
          },
          {
            "armdozer": {
              "battery": 5,
              "burst": {
                "recoverBattery": 3,
                "type": "ContinuousAttack",
              },
              "effects": [],
              "enableBurst": true,
              "hp": 2700,
              "id": "WING_DOZER",
              "maxBattery": 5,
              "maxHp": 2700,
              "name": "ウィングドーザ",
              "power": 2200,
              "speed": 2200,
            },
            "pilot": {
              "enableSkill": true,
              "id": "TSUBASA",
              "name": "ツバサ",
              "skill": {
                "batteryEnhancement": 1,
                "duration": 2,
                "type": "BatteryEnhancementSkill",
              },
            },
            "playerId": "mock-id-2",
          },
        ],
      },
    ],
  },
  "battleConnections": [
    {
      "connectionId": "owner-connection-id",
      "state": {
        "battleID": "mock-id-3",
        "players": [
          {
            "armdozer": {
              "burst": {
                "recoverBattery": 5,
                "type": "RecoverBattery",
              },
              "id": "SHIN_BRAVER",
              "maxBattery": 5,
              "maxHp": 3100,
              "name": "シンブレイバー",
              "power": 2000,
              "speed": 2000,
            },
            "connectionId": "owner-connection-id",
            "pilot": {
              "id": "SHINYA",
              "name": "シンヤ",
              "skill": {
                "recoverBattery": 2,
                "type": "RecoverBatterySkill",
              },
            },
            "playerId": "mock-id-1",
            "userID": "owner",
          },
          {
            "armdozer": {
              "burst": {
                "recoverBattery": 3,
                "type": "ContinuousAttack",
              },
              "id": "WING_DOZER",
              "maxBattery": 5,
              "maxHp": 2700,
              "name": "ウィングドーザ",
              "power": 2200,
              "speed": 2200,
            },
            "connectionId": "user2-connection-id",
            "pilot": {
              "id": "TSUBASA",
              "name": "ツバサ",
              "skill": {
                "batteryEnhancement": 1,
                "duration": 2,
                "type": "BatteryEnhancementSkill",
              },
            },
            "playerId": "mock-id-2",
            "userID": "user2",
          },
        ],
        "type": "InBattle",
      },
      "userID": "owner",
    },
    {
      "connectionId": "user2-connection-id",
      "state": {
        "battleID": "mock-id-3",
        "players": [
          {
            "armdozer": {
              "burst": {
                "recoverBattery": 5,
                "type": "RecoverBattery",
              },
              "id": "SHIN_BRAVER",
              "maxBattery": 5,
              "maxHp": 3100,
              "name": "シンブレイバー",
              "power": 2000,
              "speed": 2000,
            },
            "connectionId": "owner-connection-id",
            "pilot": {
              "id": "SHINYA",
              "name": "シンヤ",
              "skill": {
                "recoverBattery": 2,
                "type": "RecoverBatterySkill",
              },
            },
            "playerId": "mock-id-1",
            "userID": "owner",
          },
          {
            "armdozer": {
              "burst": {
                "recoverBattery": 3,
                "type": "ContinuousAttack",
              },
              "id": "WING_DOZER",
              "maxBattery": 5,
              "maxHp": 2700,
              "name": "ウィングドーザ",
              "power": 2200,
              "speed": 2200,
            },
            "connectionId": "user2-connection-id",
            "pilot": {
              "id": "TSUBASA",
              "name": "ツバサ",
              "skill": {
                "batteryEnhancement": 1,
                "duration": 2,
                "type": "BatteryEnhancementSkill",
              },
            },
            "playerId": "mock-id-2",
            "userID": "user2",
          },
        ],
        "type": "InBattle",
      },
      "userID": "user2",
    },
  ],
  "notChosenConnections": [
    {
      "connectionId": "user1-connection-id",
      "state": {
        "type": "None",
      },
      "userID": "user1",
    },
    {
      "connectionId": "user3-connection-id",
      "state": {
        "type": "None",
      },
      "userID": "user3",
    },
  ],
}
`;
